### PR TITLE
Run various queries from other queries instead of explicitly in phases

### PR DESCRIPTION
**Pull request by oli-obk**
These are just legacy leftovers from when rustc didn't have a query system. While there are more cleanups of this sort that can be done here, I want to land them in smaller steps.

This phased order of query invocations was already a lie, as any query that looks at types (e.g. the wf checks run before) can invoke e.g. const eval which invokes borrowck, which invokes typeck, ...

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/mod.rs [hunk 1]
```diff
@@ -71,6 +72,16 @@ fn mir_build(tcx: TyCtxt<'_>, def: ty::WithOptConstParam<LocalDefId>) -> Body<'_
             // of `mir_build`, so now we can steal it
             let thir = thir.steal();
 
+            if def.const_param_did.is_none() {
+                parallel!({ tcx.ensure().check_match(def.did) }, {
```
Comment (Zoxc) [inline compiler/rustc_mir_build/src/build/mod.rs:76]: I would avoid parallelism deeper in the query system to avoid the deadlock risk that brings.

File: compiler/rustc_codegen_cranelift/scripts/filter_profile.rs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env bash
-#![forbid(unsafe_code)]/* This line is ignored by bash
+#![forbid(unsafe_code)]
+/* This line is ignored by bash
```
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/scripts/filter_profile.rs:3]: Please revert this change. It is necessary for being both valid rust code and valid bash script.

File: compiler/rustc_hir_analysis/src/lib.rs [hunk 3]
```diff
@@ -529,8 +529,6 @@ pub fn check_crate(tcx: TyCtxt<'_>) -> Result<(), ErrorGuaranteed> {
         tcx.hir().for_each_module(|module| tcx.ensure().check_mod_item_types(module))
     });
 
-    tcx.sess.time("item_bodies_checking", || tcx.typeck_item_bodies(()));
```
Comment (bjorn3) [inline compiler/rustc_hir_analysis/src/lib.rs:532]: What is going to ensure that typecheck errors for all functions are reported without this call?

Comment (bjorn3) [reply compiler/rustc_hir_analysis/src/lib.rs:532]: This also regresses the quality of the output of `-Ztime-passes`.

File: compiler/rustc_codegen_cranelift/scripts/filter_profile.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_codegen_cranelift/scripts/filter_profile.rs:3]: oh huh... rustfmt changed this.

File: compiler/rustc_hir_analysis/src/lib.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_hir_analysis/src/lib.rs:532]: as mentioned in the main post

> This phased order of query invocations was already a lie, as any query that looks at types (e.g. the wf checks run before) can cause e.g. const eval which invokes borrowck, which invokes typeck, ...

we already didn't produce correct time tracking for it

File: compiler/rustc_codegen_cranelift/scripts/filter_profile.rs (hunk 2 above)
Comment (bjorn3) [reply compiler/rustc_codegen_cranelift/scripts/filter_profile.rs:3]: It is in the ignore section of `rustfmt.toml`, but rust-analyzer doesn't respect this.

**Comment by jackh726** [comment]
I think this is a good change, but maybe a safer intermediate step would be to "assert" that all the queries were run in the "passes" for now. This would ensure that we don't accidentally make a query more lazy, then skip something (accidentally).

**Comment by oli-obk** [comment]
Hmm... we don't have the setup for doing that (also doing it at that phase won't work, as it may get run later). I'll see if I can add something to the query system for asserting that a query has been run

**Comment by oli-obk** [comment]
> This would ensure that we don't accidentally make a query more lazy, then skip something (accidentally).

I looked into it, and we're already relying on the `mir_built` query to run `thir_check_unsafety`. So we're already assuming that `mir_built` gets run for all things with bodies. I agree some assertions would be neat, but generating a check-function per query and calling it is a lot. Limiting it to certain queries can be done with more query flags, but that seems like a lot of infrastructure for something that we can be certain won't happen (as we'd already be missing checks in that case).

**Comment by oli-obk** [comment]
@rust-timer build 41c8d3a20cad3d4de552339500551ada93f0d722

**Comment by cjgillot** [comment]
I'm not sure I agree with the general direction. I'd rather suggest doing the opposite:
- remove useless dependencies from queries -> avoid useless recomputations;
- invoke all the required "check" queries from `analysis` -> ensure correctness-related stuff are in a single place.

The fact that we run `thir_check_unsafety` from `mir_built` is more accidental, due to the fact that `mir_built` steals THIR. We probably shouldn't really rely on it for correctness. Even if having `ensure().mir_borrowck` in `analysis` is enough to make sure that `mir_built` is computed and `thir_check_unsafety` too.

**Comment by oli-obk** [comment]
The problem with those useless dependencies is that e.g. const eval can run on code that doesn't satisfy various checks, leading to surprising bugs and workarounds in const eval. I agree that mir_built may be too early, but at the very least the mir_drops_elaborated_and_const_checked query should not finish before the various checks on the same def id are done.

I guess my assumption was that at some point all we have to do is to either call the optimized_mir query for the main function and can compile from just the information from that, or call the analysis query to do a check build.

If changing any of this negatively affects incremental, then we should not do it, but I don't see that being an issue in general with this scheme, just with the details of where to add edges.

**Comment by oli-obk** [comment]
Everything but `assoc_many_items` in `doc` mode is noise.

**Comment by oli-obk** [comment]
> * remove useless dependencies from queries -> avoid useless recomputations;

about that, do we really add more dependencies with `ensure`? Doesn't `ensure` just mean that the ensured query must be run before the current one without any change in the dep-graph otherwise?

**Comment by cjgillot** [comment]
Yes, `ensure` adds a dependency edge. And the calls to `ensure` just before stealing a value bother me a bit: they introduce a dependency to a value that will not be used, and don't even make sure the code will not be called later. `ensure` may "just" mark the query as green, without loading any value, to be recomputed later when the value is actually required.

> The problem with those useless dependencies is that e.g. const eval can run on code that doesn't satisfy various checks, leading to surprising bugs and workarounds in const eval. I agree that mir_built may be too early, but at the very least the mir_drops_elaborated_and_const_checked query should not finish before the various checks on the same def id are done.

In those cases, we should rely on the `trainted_by_errors` fields on typeck and MIR body.

One other symptom is the calls to `track_errors` in `rustc_hir_analysis::check_crate` that shouldn't exist: typeck should rely on propagated error state via `ty::Error` and `ty::ReError`. But that FIXME exists since 2020, so that's not high priority either.

> If changing any of this negatively affects incremental, then we should not do it, but I don't see that being an issue in general with this scheme, just with the details of where to add edges.

From the perf results, I have no proof that it does negatively affect incremental, so no serious grounds on which to block this PR. That's also why I frame it as a design question, more than an operational question.

That said, we should merge that PR, as it does not lock anything in a way or the other. I'll have to make my case with an actual PR demonstrating the direction I have in mind :smile:.

**Comment by oli-obk** [comment]
> In those cases, we should rely on the `trainted_by_errors` fields on typeck and MIR body.

We do, but unsafety checks run in a query, and now, that query is actually invoked before const eval runs. Before this PR that was not guaranteed.

**Comment by oli-obk** [comment]
> That said, we should merge that PR, as it does not lock anything in a way or the other. I'll have to make my case with an actual PR demonstrating the direction I have in mind 😄.

Well, this PR isn't in a rush, and I don't want to make things harder for you. If we land this one we'll land more afterwards. Could this cause a soft lock-in that will make it hard/impossible for you to later change it to reduce the number of dependency edges?

**Comment by cjgillot** [comment]
Seeing https://github.com/rust-lang/rust/pull/108705, I finally convinced myself that this PR is the right direction. This can help to avoid building MIR for broken code, which is that much potential ICEs and non-sensical diagnostics.

> https://github.com/rust-lang/rust/pull/108118#issuecomment-1433083909
> Hmm... we don't have the setup for doing that (also doing it at that phase won't work, as it may get run later). I'll see if I can add something to the query system for asserting that a query has been run

That's what `ensure()` does: either the query is known green and they replay diagnostics from the previous compilation, or the query is re-executed.

**Comment by oli-obk** [comment]
> That's what `ensure()` does: either the query is known green and they replay diagnostics from the previous compilation, or the query is re-executed.

Oh I meant something that asserts it instead of executing it. Though ensuring later will ICE due to stolen stuff, so it results in the same thing

**Review by wesleywiser — APPROVED** [review]
